lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Solve.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE LambdaCase #-}
+
+module LambdaBuffers.Compiler.TypeClass.Solve (solve) where
+
+import Data.List (foldl', sortBy)
+import Data.Text (Text)
+import LambdaBuffers.Compiler.TypeClass.Pat (
+  Pat (AppP, DecP, ProdP, RecP, RefP, SumP, VarP, (:*), (:=)),
+  matches,
+ )
+import LambdaBuffers.Compiler.TypeClass.Rules (
+  Class (csupers),
+  Constraint (C),
+  Rule ((:<=)),
+  mapPat,
+  ruleHeadClass,
+  ruleHeadPat,
+ )
+
+import Data.Set qualified as S
+
+{- Variable substitution. Given a string that represents a variable name,
+   and a type to instantiate variables with that name to, performs the
+   instantiation
+-}
+subV :: Text -> Pat -> Pat -> Pat
+subV varNm t = \case
+  var@(VarP v) -> if v == varNm then t else var
+  x :* xs -> subV varNm t x :* subV varNm t xs
+  l := x -> subV varNm t l := subV varNm t x
+  ProdP xs -> ProdP (subV varNm t xs)
+  RecP xs -> RecP (subV varNm t xs)
+  SumP xs -> SumP (subV varNm t xs)
+  AppP t1 t2 -> AppP (subV varNm t t1) (subV varNm t t2)
+  RefP n x -> RefP (subV varNm t n) (subV varNm t x)
+  DecP a b c -> DecP (subV varNm t a) (subV varNm t b) (subV varNm t c)
+  other -> other
```

Comment:
Imo this seems very odd. Variable substitution should ONLY happen when you perform a type application. I'm not convinced this is not a hack of some kind...

Reply:
See some of my remarks in other comments. I made a mistake. We need `TyVarP`. `VarP` should only represent unification variables. This isn't substitution on `TyVar`s, it's substitution on "unification vars" (which is a bad name, they're really  Pattern Variables, or Template Variables or something if you like)

This is 100% not a hack; I lifted the basic concept straight from the pattern matching machinery in Haskell (ask me for a link to the section of the Haskell report if you want it).

Reply:
It's pattern matching substitution. It has its own rules. It's something extra you have* to graft onto the lambda calculus system. C.f. [here, esp 3.17.2](https://www.haskell.org/onlinereport/haskell2010/haskellch3.html#x8-580003.17.2)

* Well sorta, I read a paper once about a LC system where patterns aren't something extra grafted on, but I don't remember how it works.

Reply:
Just to be clear, when I point at Haskell internals here, it's because Haskell is _extensively studied and peer reviewed_ and we are _extremely unlikely to successfully reinvent all of the wheels_ that are relevant in this context.

Reply:
Assuming we reached an understanding on this important/relevant parts of this